src/partition/ProvidedPartition.cpp

```diff
@@ -22,122 +24,274 @@ void ProvidedPartition::communicate()
 {
   PRECICE_TRACE();
 
-  if (_m2ns.size() > 0) { // if there is no connected participant we also don't need to gather the mesh
-    Event e1("partition.gatherMesh." + _mesh->getName(), precice::syncMode);
+  prepare();
 
-    // Temporary globalMesh such that the master also keeps his local mesh
-    mesh::Mesh globalMesh(_mesh->getName(), _mesh->getDimensions(), _mesh->isFlipNormals(), mesh::Mesh::MESH_ID_UNDEFINED);
+  if (_m2ns.empty())
+    return;
 
-    if (not utils::MasterSlave::isSlave()) {
-      globalMesh.addMesh(*_mesh); // Add local master mesh to global mesh
-    }
+  // Temporary globalMesh such that the master also keeps his local mesh
+  mesh::Mesh globalMesh(_mesh->getName(), _mesh->getDimensions(), _mesh->isFlipNormals(), mesh::Mesh::MESH_ID_UNDEFINED);
+  bool       hasMeshBeenGathered = false;
 
-    // Gather Mesh
-    PRECICE_INFO("Gather mesh " + _mesh->getName());
-    if (utils::MasterSlave::isSlave()) {
-      com::CommunicateMesh(utils::MasterSlave::_communication).sendMesh(*_mesh, 0);
-    }
-    if (utils::MasterSlave::isMaster()) {
-      PRECICE_ASSERT(utils::MasterSlave::getRank() == 0);
-      PRECICE_ASSERT(utils::MasterSlave::getSize() > 1);
+  bool twoLevelInitAlreadyUsed = false;
 
-      for (int rankSlave = 1; rankSlave < utils::MasterSlave::getSize(); rankSlave++) {
-        com::CommunicateMesh(utils::MasterSlave::_communication).receiveMesh(globalMesh, rankSlave);
-        PRECICE_DEBUG("Received sub-mesh, from slave: " << rankSlave << ", global vertexCount: " << globalMesh.vertices().size());
-      }
-    }
+  for (auto m2n : _m2ns) {
```

Comment:
```suggestion
  for (auto& m2n : _m2ns) {
```
Otherwise, the m2n pointer gets copied on each iteration.